nautobot/extras/datasources/registry.py

```diff
@@ -48,20 +45,26 @@ def refresh_datasource_content(model_name, record, request, job_result, delete=F
                     job_result.log(
                         f"Error while refreshing {entry.name}: {exc}", level_choice=LogLevelChoices.LOG_FAILURE
                     )
-                    job_result.set_status(JobResultStatusChoices.STATUS_FAILURE)
-                job_result.save()
-            job_result.log(f"Data refresh from {record} complete!", level_choice=LogLevelChoices.LOG_INFO)
-            job_result.save()
+                    raise
     else:
-        # TODO(jathan): Review how we can eliminate these deliberate `set_status()` and `save()`
-        # calls,moving them to Celery-friendly primitives instead.
         for entry in get_datasource_contents(model_name):
             job_result.log(f"Refreshing {entry.name}...", level_choice=LogLevelChoices.LOG_INFO)
             try:
                 entry.callback(record, job_result, delete=delete)
             except Exception as exc:
                 job_result.log(f"Error while refreshing {entry.name}: {exc}", level_choice=LogLevelChoices.LOG_FAILURE)
-                job_result.set_status(JobResultStatusChoices.STATUS_FAILURE)
-            job_result.save()
-        job_result.log(f"Data refresh from {record} complete!", level_choice=LogLevelChoices.LOG_INFO)
-        job_result.save()
+                raise
+
+    # Now that any exception will fail a Job and Git Repository syncs are jobs,
+    # and we also cannot micro-manage the JobResult state, we had to add a final
+    # check here to ensure that any failed log events by the various content
+    # callbacks will result in this task "failing successfully" by raising an
+    # exception.
+    failure_logs = job_result.job_log_entries.filter(log_level=LogLevelChoices.LOG_FAILURE)
+    if failure_logs.exists():
+        msg = f"Failed to refresh data provided by {record}. Please see traceback."
+        job_result.log(msg, level_choice=LogLevelChoices.LOG_FAILURE)
+        raise RuntimeError(msg)
+
+    # Otherwise, log a friendly info message.
+    job_result.log(f"Data refresh from {record} complete!", level_choice=LogLevelChoices.LOG_INFO)
```

Comment:
@glennmatthews This is the block I was referencing on Slack that had to be added to maintain parity with 1.x behaviors. We can certainly change this, but to keep things moving this is the current solution.